### DISCUSSION TITLE
The Roadmap: Full-featured NVR replacement, or remain detection utility?

### DISCUSSION THREAD

**Discussion by evanheckert** [blakeblackshear/frigate#4573] [Ideas]
I'm in the midst of outlining functionality for [Frigate iOS app](https://github.com/blakeblackshear/frigate/discussions/4002) MVP. The challenge I'm running into to is this: My desire is for a great mobile app for a full-fledged NVR, and in its current state, Frigate is not a full-fledged NVR.

Note that this is *not* a criticism - nor have I seen anyone claim that it is intended to be. It does bring up a question that's important to those contributing to the ecosystem around Frigate:

**Is the medium-to-long term vision for the core Frigate project to remain focused on detection, or is the vision to become a full-featured replacement for an NVR?**

This helps answer questions like:
* Should I continue to rely on something like blue iris?
* Should I build out the other NVR features as a separate project, using Frigate as the ingest/detect engine?
* etc...

As a point-of-reference, here's a work-in-progress outline of a (subjectively) ideal NVR. I'll come back and update this as I complete more of it. Feel free to chime in with things that you would think should be included.

* Recording & Detection
    * Detection
        * Hardware acceleration
    * Schedules
        * Default schedule is Sun-Sat, 00:00-23:59
        * Can break week up into multiple named schedules
        * Rules for Detect and Record
            * Zones
    * Record & Detect Settings
        * Defined per-schedule
            * Optional Per-Camera settings
            * Optional Per-Group settings
            * Fallback all-camera settings
        * Record
            * Video On/Off
            * Audio On/Off
        * Detect
            * Triggers
                * Motion
                * Other Cameras
                * Manual Trigger
                * Sensors, etc. via MQTT
            * What objects to detect
            * Sensitivity
            * Per-camera, can set those settings per-zone
            * Snapshot options…
            * Clip Recording options…
            * Trigger other cameras/groups
    * Media Types
        * Snapshots (image)
        * Clips (short video)
        * Full Recording (long video)
    * Storage Management
        * Dedicated paths per media type
        * Storage constraints per path
        * Camera-level overrides
* Camera Control
    * Live-Stream
    * Reboot
    * Two-way audio
    * Pan/Tilt/Zoom where applicable
    * Disable audio/video per schedule and permissions
* Notifications
    * Multiple protocols, including IMAP, MQTT, push via mobile app, etc.
    * Admin notifications about
        * new releases/updates avail
        * system resource exceed threshold
        * storage approaching/exceeded thresholds
        * camera down, other errors
* Web Server & Mobile App
    * Admin
        * View/Update Settings
        * Add/Remove/Reboot Camera(s)
        * Change Camera/Group Settings
        * System status & Resource Monitoring
        * View Logs
    * User defined multi-cam live views
    * Individual Cam
        * Live-View Audio/Video
        * Control PTZ, Light, IR
        * Speak, Sound Alarm
        * Trigger 
        * Manual Record
* API
* Mobile App
* Users
    * Secure Authentication
    * Permissions, default and per-camera
    * Notification Preferences
    * Activity Logs
* Locations
    * SSO to access multiple locations

**Comment by blakeblackshear** [comment]
I think many users definition of a "full-fledged" NVR will be different. For many users, Frigate has everything they need in an NVR. I have no need to run blue iris or any other software other than Frigate for my cameras. Even still, Frigate already offers most of the features you listed.

I wouldn't say that I am targeting any arbitrary set of features, and I am certainly not just trying to check the boxes to get feature parity with what other NVR solutions offer. I would think about adding things to Frigate based on the individual merits of the features themselves. I am also not drawing any arbitrary lines that Frigate won't ever cross in terms of features.

I do think Frigate's differentiator is the way object detection is a part of the core processing pipeline rather than a bolt on afterthought. I am more likely to lean into the features that set it apart rather than into the things that emulate other NVRs out there.

**Reply by evanheckert** [reply]
> I do think Frigate's differentiator is the way object detection is a part of the core processing pipeline rather than a bolt on afterthought. I am more likely to lean into the features that set it apart rather than into the things that emulate other NVRs out there.

I'm not surprised to hear you say that, given how focused the project has stayed even whilst gaining popularity. That seems like a solid long term-play, and fortunately you have that freedom since you're not beholden to any particular mass adoption growth targets. 

Regarding merit-based feature evaluation - how is the merit evaluated? A few examples:
* the extent to which a feature contributes to its core differentiation, from the perspective of the maintainers
* the perceived net value gained by all existing users, based on github discussions, polling, etc
* the extent to which it lowers the friction of adoption for a wider audience, maximizing growth of the userbase

Of course, could be any combination of the above plus the infinite possibilities not listed, but I am curious your thought process. If helpful to discuss through an example, here are the top three that make it incomplete as a recommendation to a few local business owners and my parents:

* User/Account based permissions (without Home Assistant)
* Add/Remove/Change Camera Config from GUI
* Intuitive native mobile app with live view with PTZ control, video scrubbing, and GUI config editing, essentially feature parity to what the web app would have, but with a touch-native UI that'd be intuitive and not intimidating for non-technical people.

**Reply by blakeblackshear** [reply]
As you guessed, there is no defined rubric for evaluating feature requests. It's a combination of lots of different things and will evolve over time. Here is where my head is right this moment. 

I think the question you are really asking is whether or not frigate will become an NVR system that can be setup, managed, and used by someone who isn't at least fairly technical. I hope someday, but that is not my immediate goal. I think there are many improvements to be made in the UI features, accuracy of object detection, accuracy of object tracking, stability of the camera connection, and several other places before that becomes the primary focus. 

At the moment, the overwhelming majority of development and support is provided is provided by @NickM-27 and myself on the side of our existing jobs. This includes monitoring reddit, HA Discord, and home assistant community forums on top of the issues created here. It is tremendously useful that we can ask users who open issues to do things like exec into the container and run commands. The majority of our current users can actually do that. I honestly don't think we could keep up if we were being inundated with requests from non technical users right now. I would rather focus that energy on improving Frigate for the existing community for now and iterate our way there. 

Perhaps someday I will be able to follow in the footsteps of Nabu Casa and turn this project into something that can support full time development and support resources with actual branded hardware. I am hoping Frigate+ could possibly create a pathway to that future, but who knows. 

In terms of the specific features you listed, all of those are already pinned as things we want to add. We have described a path to config editing in the UI, discussed better UIs for video scrubbing, and ways to have a viewer mode in the UI. None of them are off the table. 

I'm trying to be as transparent and genuine as possible.
